lib/util/StackedSetMap.js

```diff
@@ -5,24 +5,72 @@
 
 "use strict";
 
-const TOMBSTONE = {};
-const UNDEFINED_MARKER = {};
+const TOMBSTONE = Symbol("tombstone");
+const UNDEFINED_MARKER = Symbol("undefined");
+
+/**
+ * @template T
+ * @typedef {T | true | undefined} Cell<T>
+ */
+
+/**
+ * @template T
+ * @typedef {T | true | typeof TOMBSTONE | typeof UNDEFINED_MARKER} InternalCell<T>
+ */
+
+/**
+ * @template K
+ * @template V
+ * @param {[K, InternalCell<V>]} pair the internal cell
+ * @returns {[K, Cell<V>]} its “safe” representation
+ */
+const extractPair = pair => {
+	const key = pair[0];
+	const val = pair[1];
+	if (val === UNDEFINED_MARKER || val === TOMBSTONE) {
+		return [key, undefined];
+	} else {
+		return [key, val];
```

Comment:
It might put pressure on GC if it's called very often. In normal TS it's possible to write `something as Type`, there should be a way to re-declare the type here and still return `pair` without creating a new array?